### PR TITLE
drivers/adc: add Gecko IADC driver

### PR DESCRIPTION
**Pull request by fkokosinski**
This PR adds the Gecko IADC driver and support for it to the efr32bg_sltb010a board.

### CODE REVIEW

**Review by tgorochowik — DISMISSED** [review]

**Comment by fkokosinski** [comment]
Hi @anangl! Could take a look at this PR?

**Review by kgugala — DISMISSED** [review]
LGTM

File: drivers/adc/iadc_gecko.c [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+		channels &= ~BIT(index);
+	}
+
+	/* Check buffer size */
+	res = adc_gecko_check_buffer_size(sequence, channel_count);
+	if (res < 0) {
+		return res;
+	}
+
+	adc_gecko_set_config(data->dev);
```
Comment (anangl) [inline drivers/adc/iadc_gecko.c:153]: Why is it called here when `data->channel_id` has not been set yet?

File: drivers/adc/iadc_gecko.c [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+		LOG_ERR("Oversampling is not supported");
+		return -ENOTSUP;
+	}
+
+	/* Check resolution setting */
+	res = adc_gecko_check_resolution(sequence);
+	if (res < 0) {
+		return -EINVAL;
+	}
+	data->resolution = res;
```
[the remark this answers is not in the document]
Comment (fkokosinski) [reply drivers/adc/iadc_gecko.c]: Thanks for spotting it. It indeed isn't used, so I've removed it.

File: dts/bindings/adc/silabs,gecko-iadc.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+compatible: "silabs,gecko-iadc"
+
+include: adc-controller.yaml
+
+properties:
+  reg:
+    required: true
+
+  interrupts:
+    required: true
```
[the remark this answers is not in the document]
Comment (fkokosinski) [reply dts/bindings/adc/silabs,gecko-iadc.yaml:15]: Done

File: drivers/adc/iadc_gecko.c [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+
+	/* Verify all requested channels are initialized and store resolution */
+	channels = sequence->channels;
```
[the remark this answers is not in the document]
Comment (fkokosinski) [reply drivers/adc/iadc_gecko.c:132]: Introduced a check here.

File: drivers/adc/iadc_gecko.c (hunk 1 above)
Comment (fkokosinski) [reply drivers/adc/iadc_gecko.c]: You're right that it shouldn't be called here. Removed.

**Comment by fkokosinski** [comment]
@anangl thanks for the review. Please let me know if anything else needs addressing/fixing. The reported regression (twister job failure) is not introduced by this PR.

**Comment by fkokosinski** [comment]
Hi @anangl! All checks except the DNM one are now green after https://github.com/zephyrproject-rtos/zephyr/pull/56806 got merged.

**Review by anangl — DISMISSED** [review]

**Comment by fkokosinski** [comment]
Hi @anangl! I had to rebase the PR due to Zephyr and HAL changes. Could you PTAL?

**Review by anangl — APPROVED** [review]

**Review by tgorochowik — APPROVED** [review]

**Pull request merged**